prov/sockets/src/sock_conn.c

```diff
@@ -213,33 +215,66 @@ int fd_set_nonblock(int fd)
 	return ret;
 }
 
-void sock_set_sockopt_reuseaddr(int sock)
+#if !defined __APPLE__ && !defined _WIN32
+void sock_set_sockopt_keepalive(int sock)
 {
 	int optval;
+
+	/* Keepalive is disabled: now leave */
+	if (!sock_keepalive_enable)
+		return;
+
 	optval = 1;
-	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
-		SOCK_LOG_ERROR("setsockopt reuseaddr failed\n");
+	if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, &optval, sizeof(optval)))
+			SOCK_LOG_ERROR("setsockopt keepalive enable failed: %s\n",
+			               strerror(errno));
+
+	if (sock_keepalive_time != INT_MAX) {
+		optval = sock_keepalive_time;
+		if (setsockopt(sock, SOL_TCP, TCP_KEEPIDLE, &optval, sizeof(optval)))
+			SOCK_LOG_ERROR("setsockopt keepalive time failed: %s\n",
+			               strerror(errno));
+	}
+
+	if (sock_keepalive_intvl != INT_MAX) {
+		optval = sock_keepalive_intvl;
+		if (setsockopt(sock, SOL_TCP, TCP_KEEPINTVL, &optval, sizeof(optval)))
+			SOCK_LOG_ERROR("setsockopt keepalive intvl failed: %s\n",
+			               strerror(errno));
+	}
+
+	if (sock_keepalive_probes != INT_MAX) {
+		optval = sock_keepalive_probes;
+		if (setsockopt(sock, SOL_TCP, TCP_KEEPCNT, &optval, sizeof(optval)))
+			SOCK_LOG_ERROR("setsockopt keepalive intvl failed: %s\n",
+			               strerror(errno));
+	}
 }
+#endif
 
-void sock_set_sockopts_conn(int sock)
+static void sock_set_sockopt_reuseaddr(int sock)
 {
 	int optval;
 	optval = 1;
-	sock_set_sockopt_reuseaddr(sock);
-	if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval)))
-		SOCK_LOG_ERROR("setsockopt nodelay failed\n");
+	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
+		SOCK_LOG_ERROR("setsockopt reuseaddr failed\n");
 }
 
-void sock_set_sockopts(int sock)
+void sock_set_sockopts(int sock, enum sock_opts sock_opts)
 {
 	int optval;
 	optval = 1;
 
 	sock_set_sockopt_reuseaddr(sock);
+#if !defined __APPLE__ && !defined _WIN32
+	if (sock_opts & SOCK_OPTS_KEEPALIVE)
+		sock_set_sockopt_keepalive(sock);
+#endif
```

Comment:
If-defs in the middle of functions make for difficult to maintain code.  Please always call and wrap the ifdef around sock_set_sockopt_keepalive() instead.